File: Boxuan/16_01_18/Copy_Books.py
class Solution:
    # @param pages: a list of integers
    # @param k: an integer
    # @return: an integer
    def copyBooks(self, pages, k):
        # write your code here
        if not pages:
            return 0
        left, right = max(pages), sum(pages)
        while left + 1 < right:
            mid = left + (right - left) / 2
            if self.couterCopiers(pages, mid) > k:
                left = mid
            else:
                right = mid
        ans = left if self.couterCopiers(pages, left) < k else right
        return ans
        
    def couterCopiers(self, pages, limit):
        if not pages:
            return 0
        copier = 1
        sum = pages[0]
        for i in range(1, len(pages)):
            if sum + pages[i] > limit:
                sum = 0
                copier += 1
            sum += pages[i]
        return copier

File: Yixuan/17_01_17/copy_books.cpp
class Solution {
public:
    /**
     * @param pages: a vector of integers
     * @param k: an integer
     * @return: an integer
     * @link: http://www.lintcode.com/en/problem/copy-books/
     * @author: Egbert Li
     */
   int copyBooks(vector<int> &pages, int k) {
      int length = pages.size();
      if (length == 0) {
         return 0;
      }
      int max = pages[0], total = 0;
      for (int i = 0; i < length; i++) {
         total += pages[i];
         if (pages[i] > max) {
            max = pages[i];
         }
      }
      int start = max;
      int end = total;
      while (start + 1 < end) {
         int mid = start + (end - start) / 2;
         if (countCopiers(pages, mid) > k) {
            start = mid;
         } else {
            end = mid;
         }
      }
      if (countCopiers(pages, start) <= k) {
         return start;
      }
      return end;
   }
// according the current pages get possible people to copy
private:
   int countCopiers(vector<int> &pages, int limit) {
      if (pages.size() == 0) {
          return 0;
      }
      int sum = pages[0];
      int copiers = 1;
      for (int i = 1; i < pages.size(); i++) {
         if (sum + pages[i] > limit) {
            sum = 0;
            copiers++;
         }
         sum += pages[i];
      }
      return copiers;
   }
};
